File: web/src/main/java/uk/co/rafearnold/randomworkoutroutine/web/service/RoutineService.java
package uk.co.rafearnold.randomworkoutroutine.web.service;

import uk.co.rafearnold.randomworkoutroutine.web.model.entity.Group;
import uk.co.rafearnold.randomworkoutroutine.web.model.entity.Routine;
import uk.co.rafearnold.randomworkoutroutine.web.repository.RoutineRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class RoutineService extends ItemService<Routine> {

    public RoutineService(RoutineRepository repository) {
        super(repository);
    }

    public void removeGroupFromAll(Group group) {
        List<Routine> routines = ((RoutineRepository) repository).findAllByGroupsContaining(group);
        for (Routine routine : routines) {
            routine.setGroups(routine.getGroups().stream()
                    .filter(g -> g.getId() != group.getId())
                    .collect(Collectors.toList()));
            repository.save(routine);
        }
    }

    @Override
    protected Routine createItem(UUID id, String name) {
        Routine routine = new Routine();
        routine.setId(id);
        routine.setName(name);
        return routine;
    }
}

File: web/src/main/kotlin/uk/co/rafearnold/randomworkoutroutine/web/service/RoutineService.kt
package uk.co.rafearnold.randomworkoutroutine.web.service

import org.springframework.stereotype.Service
import uk.co.rafearnold.randomworkoutroutine.web.model.entity.Exercise
import uk.co.rafearnold.randomworkoutroutine.web.model.entity.ExerciseOption
import uk.co.rafearnold.randomworkoutroutine.web.model.entity.Group
import uk.co.rafearnold.randomworkoutroutine.web.model.entity.Routine
import uk.co.rafearnold.randomworkoutroutine.web.repository.ExerciseOptionRepository
import uk.co.rafearnold.randomworkoutroutine.web.repository.GroupRepository
import uk.co.rafearnold.randomworkoutroutine.web.repository.RoutineRepository
import java.util.*

/**
 * Service for retrieving and manipulating [Routine] entities.
 */
@Service
class RoutineService(
    routineRepository: RoutineRepository,
    private val exerciseOptionRepository: ExerciseOptionRepository,
    private val groupRepository: GroupRepository
) : ItemService<Routine>(routineRepository) {

    /**
     * Saves all [Group]s of [item], as well as each [Group]'s [ExerciseOption]s.
     */
    override fun saveChildren(item: Routine) {
        val groups: List<Group> = item.groups
        exerciseOptionRepository.saveAll(groups.flatMap { it.exercises })
        groupRepository.saveAll(groups)
    }

    /**
     * Deletes every [ExerciseOption] associated with [exercise] (i.e. any [ExerciseOption] whose
     * [ExerciseOption.exercise] property equals [exercise]). This also removes each deleted
     * [ExerciseOption] from the [Group] that contains it.
     */
    fun deleteOptionsByExercise(exercise: Exercise) {
        val exerciseOptions: Iterable<ExerciseOption> = exerciseOptionRepository.findAllByExercise(exercise)
        removeAllExerciseOptionsFromGroups(exerciseOptions)
        exerciseOptionRepository.deleteAll(exerciseOptions)
    }

    /**
     * Removes each [ExerciseOption] of [exerciseOptions] from the [Group] that contains it (i.e.
     * the [Group] whose [Group.exercises] property contains it).
     */
    private fun removeAllExerciseOptionsFromGroups(exerciseOptions: Iterable<ExerciseOption>) {
        val groups: Set<Group> =
            exerciseOptions.flatMap { groupRepository.findAllByExercisesContaining(it) }.toSet()
        val exerciseOptionIds: List<UUID> = exerciseOptions.map { it.id }
        for (group in groups) {
            group.exercises = group.exercises.filter { it.id !in exerciseOptionIds }.toMutableList()
        }
        groupRepository.saveAll(groups)
    }

    /**
     * Deletes [item], as well as its children.
     */
    override fun delete(item: Routine) {
        super.delete(item)
        val groups: List<Group> = item.groups
        groupRepository.deleteAll(groups)
        exerciseOptionRepository.deleteAll(groups.flatMap { it.exercises })
    }
}
